Support CWL workflow restarts (resolves #1241, resolves #1100)

chapmanb opened the pull request:
Continues #1025

Support restarting a cwltoil workflow with the `--restart` flag,
updates work by @tetron in #1025 and fixes #1100. Adds a test
to demonstrate restart capability by manipulating paths to cause the
initial run of the revsort workflow to fail.

Still a work in progress as the test fails consistently when restarting.
I could use help debugging what I'm missing to correctly restart to
finish this. If you run the test:

```
make test tests=src/toil/test/cwl/cwlTest.py::CWLTest::test_restart
```

It completes the expected failure step, but the restart fails with an
EOFError when unpickling a job store (similar to #676 and #1039):

```
INFO toil.jobStores.fileJobStore: Path to job store directory is '/tmp/user/1001/toil-test-toil.test.cwl.cwlTest.CWLTest-test_restart-l0vC5G/jobStore'.
INFO toil.common: Using the single machine batch system
WARNING toil.batchSystems.singleMachine: Limiting maxCores to CPU count of system (4).
WARNING toil.batchSystems.singleMachine: Limiting maxMemory to physically available memory (7916355584).
INFO toil.common: Created the workflow directory at /tmp/user/1001/toil-c7c91877-c1bd-4a6e-b2c8-b23d0c3d90b7
WARNING toil.batchSystems.singleMachine: Limiting maxDisk to physically available disk (15848153088).
INFO toil.batchSystems.singleMachine: Setting up the thread pool with 40 workers, given a minimum CPU fraction of 0.100000 and a maximum CPU value of 4.
INFO toil.common: Written the environment for the jobs to the environment file
INFO toil.common: Caching all jobs in job store
INFO toil.common: 4 jobs downloaded.
INFO toil.jobStores.abstractJobStore: Checking job graph connectivity...
INFO toil.jobStores.abstractJobStore: 4 jobs reachable from root.
CRITICAL toil.jobStores.abstractJobStore: Repairing job: N/g/jobLYOvLb
INFO toil.jobStores.abstractJobStore: Discarding old statistics and logs...
INFO toil.jobStores.abstractJobStore: Job store is clean
INFO toil.realtimeLogger: Real-time logging disabled
INFO toil.leader: (Re)building internal scheduler state
INFO toil.leader: Checked batch system has no running jobs and no updated jobs
INFO toil.leader: Found 2 jobs to start and 1 jobs with successors to run
INFO toil.leader: Starting the main loop
INFO toil.batchSystems.singleMachine: Executing command: '/home/chapmanb/bio/toil/venv/bin/_toil_worker /tmp/user/1001/toil-test-toil.test.cwl.cwlTest.CWLTest-test_restart-l0vC5G/jobStore N/g/jobLYOvLb'.
INFO toil.batchSystems.singleMachine: Executing command: '/home/chapmanb/bio/toil/venv/bin/_toil_worker /tmp/user/1001/toil-test-toil.test.cwl.cwlTest.CWLTest-test_restart-l0vC5G/jobStore N/2/jobv9Osdj'.
toil.leader: The jobWrapper seems to have left a log file, indicating failure: N/2/jobv9Osdj
toil.leader: Reporting file: N/2/jobv9Osdj
toil.leader: N/2/jobv9Osdj:   ---TOIL WORKER OUTPUT LOG---
toil.leader: N/2/jobv9Osdj:   INFO:rdflib:RDFLib Version: 4.2.1
toil.leader: N/2/jobv9Osdj:   INFO:toil.jobStores.fileJobStore:Path to job store directory is '/tmp/user/1001/toil-test-toil.test.cwl.cwlTest.CWLTest-test_restart-l0vC5G/jobStore'.
toil.leader: N/2/jobv9Osdj:   Traceback (most recent call last):
toil.leader: N/2/jobv9Osdj:     File "/home/chapmanb/bio/toil/venv/local/lib/python2.7/site-packages/toil-3.5.0a1-py2.7.egg/toil/worker.py", line 316, in main
toil.leader: N/2/jobv9Osdj:       job = Job._loadJob(jobWrapper.command, jobStore)
toil.leader: N/2/jobv9Osdj:     File "/home/chapmanb/bio/toil/venv/local/lib/python2.7/site-packages/toil-3.5.0a1-py2.7.egg/toil/job.py", line 673, in _loadJob
toil.leader: N/2/jobv9Osdj:       return cls._unpickle(userModule, fileHandle)
toil.leader: N/2/jobv9Osdj:     File "/home/chapmanb/bio/toil/venv/local/lib/python2.7/site-packages/toil-3.5.0a1-py2.7.egg/toil/job.py", line 696, in _unpickle
toil.leader: N/2/jobv9Osdj:       return unpickler.load()
toil.leader: N/2/jobv9Osdj:   EOFError
toil.leader: N/2/jobv9Osdj:   Exiting the worker because of a failed jobWrapper on host kunkel
toil.leader: N/2/jobv9Osdj:   ERROR:toil.worker:Exiting the worker because of a failed jobWrapper on host kunkel
toil.leader: N/2/jobv9Osdj:   WARNING:toil.jobWrapper:Due to failure we are reducing the remaining retry count of job N/2/jobv9Osdj to 0
toil.leader: N/2/jobv9Osdj:   WARNING:toil.jobWrapper:We have increased the default memory of the failed job to 2147483648 bytes
toil.leader: Job: N/2/jobv9Osdj is completely failed
```

Thanks for any tips and ideas about how best to debug and proceed.

cgcloud-jenkins commented:
Can one of the admins verify this patch?

cgcloud-jenkins commented:
Can one of the admins verify this patch?

arkal commented:
Jenkins, test this please.

benedictpaten commented:
Jenkins, test please

benedictpaten commented:
Jenkins, test this please.

cket commented:
The only failing test is CWL test_restart, you can see the test output here http://pastebin.com/UPqgjzLC.
It looks like these are the most relevant lines:
```
WARNING:toil.leader:X/y/jobxPa_0d:  INFO:cwltool:[job revtool.cwl] /mnt/ephemeral/tmp/tmpRYIG2q$ rev \
WARNING:toil.leader:X/y/jobxPa_0d:      /mnt/ephemeral/tmp/tmpCYsUeI/stg49ff26ed-3fbf-41ba-85bc-eac77910a205/whale.txt > /mnt/ephemeral/tmp/tmpRYIG2q/output.txt
WARNING:toil.leader:X/y/jobxPa_0d:  'rev' not found
WARNING:toil.leader:X/y/jobxPa_0d:  ERROR:cwltool:'rev' not found
```

chapmanb commented:
Christopher -- thanks for looking at this one. That's right on, that's the new test I'm adding for validating restarts and I can't quite get things working for the restart so it still fails. The initial message in the PR has more details about where I'm stuck at -- the restart complains when unpickling a file and I'm not sure what needs to happen to avoid that problem and restart cleanly. Thanks for any pointers/ideas.

cket commented:
I see now, sorry about that! I've been walking over this and I think I'm getting closer. The job with the EOF error is actually run again despite 'failing completely', for some reason. The job is only run once when I run the pipeline with the forced error failing, so this points to some problem in the restart logic causing a duplicate job being issued.  

More evidence of a scheduling error is that the job is run with the same command and unpickles the file successfully on the second try, whereas the file doesn't exist on the first try. It seems like the failing job was just run too early. My next step is identifying where exactly the job is being issued 2X.

chapmanb commented:
Christopher -- thanks for the help debugging. The test job I wrote for this should submit the job multiple times:

1. First I manipulate the PATH to make the job fail (by moving `rev`, needed for the job to run, out of the PATH).
2. Then I fix the PATH and rerun the same job with `--restart` so it'll hopefully finish cleanly.
3. Then I run again with `--restart` to test for an expected failure (restarting on a finished job).

I'm not sure if that's relevant to what you're seeing or not but is the logic behind the tests. From the new code side, I'm now calling toil.restart() on the old directory instead of doing a new workflow. Perhaps there should be more tweaking/fixes from my side, I'm just not sure what to do. Thanks again for the help.

cket commented:
Hi Brad, the problem here seems to be a job being issued multiple times within the second step of your test when the workflow is called with ```--restart```. Since the workflow runs correctly without forcing the initial failure I think this is a ```--restart`` problem, not a workflow or test problem.

When a workflow is running Toil determines what job to run next by looking a finished job's successors. In the ```--restart``` case we are seeing the EOF job being issued despite not being part of any finishing job's successors. The job fails, and is then being issued again (correctly this time) when its predecessor finishes. 

Unfortunately, I'm out of the office this week and the next so I can't dedicate too much more time to debugging but I would look at the restart code to figure out how it determines which jobs are OK to run when restarting the workflow.  Does the CWL support implement its own restart logic? If so, that's where I would looks first.

hannes-ucsc commented on src/toil/test/cwl/cwlTest.py:
```diff
@@ -50,6 +50,35 @@ def test_run_revsort(self):
+        # Force a failure and half finished job
+        try:
+            cwltoil.main(cmd)
+            raise ValueError("Expected problem job did not fail")
```
Use self.fail()

hannes-ucsc commented on src/toil/test/cwl/cwlTest.py:
```diff
@@ -50,6 +50,35 @@ def test_run_revsort(self):
+            cwltoil.main(cmd)
+            raise ValueError("Expected problem job did not fail")
+        except AssertionError:
+            pass
```
Why are you swallowing AssertionError here?

hannes-ucsc commented on src/toil/test/cwl/cwlTest.py:
```diff
@@ -50,6 +50,35 @@ def test_run_revsort(self):
+        # Should fail because job completed successfully
+        try:
+            cwltoil.main(cmd + ["--restart"])
+            raise AssertionError("Restart with missing directory did not fail")
```
self.fail("…")

hannes-ucsc commented on src/toil/test/cwl/cwlTest.py:
```diff
@@ -50,6 +50,35 @@ def test_run_revsort(self):
+            cwltoil.main(cmd + ["--restart"])
+            raise AssertionError("Restart with missing directory did not fail")
+        except NoSuchJobStoreException:
+            pass
```
Document why this exception is expected.

hannes-ucsc commented on src/toil/test/cwl/cwlTest.py:
```diff
@@ -50,6 +50,35 @@ def test_run_revsort(self):
+        os.environ["PATH"] = path_without_rev()
+        # Force a failure and half finished job
+        try:
+            cwltoil.main(cmd)
```
I don't see where the failure is being forced.

hannes-ucsc replied:
Read your PR description just now which answers my question. But I think a comment would help here.

hannes-ucsc commented:
You might want to pass --logDebug to the test. Maybe that will reveal something.

chapmanb commented:
Hannes;
Thanks for all the comments. While responding to these I realized the underlying issue might be that cwltoil is exiting poorly for this failure case, and I moved discussion of this to #1122 with more details. I hope if we fix that it'll help here. Would you have any pointers about how best to improve failures with cwltoil runs so they're cleaner?

benedictpaten commented:
@chapmanb @cket I think this might now work, I fixed the big bug in toil that was preventing it working correctly. Hope this can be unblocked.

chapmanb commented:
Benedict -- thanks so much for the fixes. This does resolve the problem and get this working correctly. Brilliant. I've updated the tests to synchronize with the latest changes and address Hannes' review and rebased this all into a single clean commit.

So, this looks great to me now and is ready to test/merge if anyone on the Toil team had time to review. Thanks again for all the help getting this working.

hannes-ucsc commented:
IT failed due to PyPI outage.
